### PR TITLE
Various mypy related updates

### PR DESCRIPTION
**Pull request by phlogistonjohn**
While working on one task I ran into some mypy related issue that I thought may be fixed in newer versions of mypy. But I was not easily able to try a newer version because the version is pinned in a "global" mypy constraints file - and updating the version there caused various new issues to appear.

After a period of groaning I decided that trying to make the codebase work on newer versions of mypy was worthwhile and something I had thought of doing previously anyway. I've tested on versions of mypy from 0.990 - 1.8.0. This PR *does not* change the pinned version of mypy it just makes it easier to make that change in the future.

Most everything the newer versions of mypy found were valid, with the exception of the detecting the cython based extension sources as empty namespace packages forcing me the turn off namespace package support. However, this seems to have had no negative side effects so it seems like a valid workaround for now.

One change (currently prefixed by `xxx` in the commit message) in the rook module seems to have a legit logical error. But I wasn't clear if my fix is correct. So I may need some help verifying it before taking the PR out of draft.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
  - [x] Touches a number of different components, so you tell me if you really really want a ticket.
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests
  - [x] This is not a feature. Existing tests ought to be sufficient.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
@rkachach can you please take a look into the final commit in this PR (`xxx: NOT VERIFIED: pybing/mgr/rook: change logic to prevent impossibl…`). Because while I think I corrected an error I don't really know how to verify it. Let me know if it's OK or if we have a way to test it. I'm keeping the PR in draft until I hear from someone who is familiar enough with the rook module to give the patch a thumbs -up. Thanks!

File: src/pybind/mgr/rook/rook_cluster.py [hunk 1]
```diff
@@ -444,7 +444,7 @@ def filter_devices(self, rook_pods: KubernetesResource, drive_group: DriveGroupS
                             all 
                             or (
                                 device.sys_api['node'] in matching_hosts
-                                and ((sizematcher != None) or sizematcher.compare(device))
+                                and ((sizematcher is None) or sizematcher.compare(device))
```
Comment (tchaikov) [inline src/pybind/mgr/rook/rook_cluster.py:447]: i am far from an expert in rook mgr module, but i think the fix is correct.  and probably a better fix would be to initialize `sizematcher` with `AllMatcher('', None)` instead, like
```py
sizematcher: Matcher = AllMatcher('', None)
# ...
all 
or (device.sys_api['node'] in matching_hosts
    and sizematcher.compare(device)
    and ...
)
```

Comment (phlogistonjohn) [reply src/pybind/mgr/rook/rook_cluster.py]: Thanks, that suggestion is much clearer to read. I have pushed a change for that. I'll still wait for  a general thumbs-up/down from someone with rook module experience, at least for a day or two.

Comment (rkachach) [reply src/pybind/mgr/rook/rook_cluster.py]: @phlogistonjohn I think both your change and the improvement suggested by @tchaikov are correct. Anyway, as far as I know this code is not being used in rook latest versions anymore (it's part of OSDs creation which is now managed by the Rook orch).

**Review by rkachach — APPROVED** [review]
rook related changes LGTM

**Comment by tchaikov** [comment]
hi John, thank you for fixing the mypy issues! mypy did help us to address quite a few issues lurking around in the mgr modules when we introduced the type annotation system to the source tree. i am very glad it is still very helpful with your efforts.

> This PR does not change the pinned version of mypy it just makes it easier to make that change in the future.

so, IIUC, there are still some issues identified by newer version of mypy even after your change, right?

**Comment by phlogistonjohn** [comment]
All required checks are passing... moving to ready for review.

**Comment by phlogistonjohn** [comment]
> so, IIUC, there are still some issues identified by newer version of mypy even after your change, right?

Unless I overlooked something, no - the codebase should be clean for all currently released versions of mypy. However, not all versions of mypy work with all the python versions ceph currently supports. Plus, the way the tox.ini files are written in the ceph tree makes it difficult to try different combinations of mypy / python. On top of that I wanted to research if we could use the [python version option](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-python-version). And I was hesitant to change it before the work getting centos 9 working was done. So I chose to wait and make the version bump in a later pr.

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -1410,7 +1410,7 @@ def from_json(cls, data: dict) -> 'ServiceDescription':
         return cls(spec=spec, events=events, **c_status)
 
     @staticmethod
-    def yaml_representer(dumper: 'yaml.SafeDumper', data: 'ServiceDescription') -> Any:
+    def yaml_representer(dumper: 'yaml.Dumper', data: 'ServiceDescription') -> yaml.Node:
```
Comment (adk3798) [inline src/pybind/mgr/orchestrator/_interface.py:1413]: just curious, did mypy actually complain about the use of `Any` here, or did you just want to make the type more specific

**Review by adk3798 — APPROVED** [review]
cephadm/orchestrator pieces LGTM

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (phlogistonjohn) [reply src/pybind/mgr/orchestrator/_interface.py:1413]: It complained. But after looking into it again, it's not merely a new version of mypy alone but a newer version of `types-PyYAML` is also needed to get the errors:
```
orchestrator/__init__.py:6: note: In module imported here:
orchestrator/_interface.py:1284: error: Argument 2 to "add_representer" has incompatible type "Callable[[SafeDumper, DaemonDescription], Any]"; expected "Callable[[Dumper, DaemonDescription], Node]"  [arg-type]
orchestrator/_interface.py:1417: error: Argument 2 to "add_representer" has incompatible type "Callable[[SafeDumper, ServiceDescription], Any]"; expected "Callable[[Dumper, ServiceDescription], Node]"  [arg-type]
orchestrator/__init__.py:3: note: In module imported here:
orchestrator/module.py:136: error: Argument 2 to "add_representer" has incompatible type "Callable[[SafeDumper, HostDetails], Any]"; expected "Callable[[Dumper, HostDetails], Node]"  [arg-type]

```

File: qa/workunits/windows/test_rbd_wnbd.py [hunk 3]
```diff
@@ -551,7 +551,7 @@ class RbdFioTest(RbdTest):
 
     def __init__(self,
                  *args,
-                 fio_size_mb: int = None,
+                 fio_size_mb: int = 0,
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply qa/workunits/windows/test_rbd_wnbd.py:554]: Done. I wasn't sure if you preferred to have just the default value changed or changed + type annotated so I did the latter. I can undo it if that's an issue.

**Review by Pegonzal — APPROVED** [review]
the mgr/dashboard changes look good as well

**Comment by phlogistonjohn** [comment]
@adk3798 I think we have sufficient reviews from the various functional areas to look towards merging. Let me know if there's anything else that needs to be handled before then.

**Pull request merged**